crates/fuel-core/src/query/balance/asset_query.rs

```diff
@@ -120,6 +123,11 @@ impl<'a> AssetsQuery<'a> {
                 }
             });
 
+        let predicate = self
+            .assets
+            .as_ref()
+            .map(|assets| assets.contains(self.base_asset_id))
+            .unwrap_or(true);
```

Comment:
nit: If this evaluates to `false`, we should do a short-circuiting return somehow. Sadly Rust makes this vary hard, but maybe moving the following iterator chain to a helper function would do it.

In any case I don't think this should be named `predicate`, when it's in fact a boolean. Consider something like `has_base_asset`.

Reply:
Already much more readable. I dislike the `Box` here, but going halfway with

```rust
    pub fn coins(&self) -> impl Iterator<Item = StorageResult<CoinType>> + '_ {
        let has_base_asset = self.has_asset(self.base_asset_id);
        self.coins_iter().chain(self.messages_iter().take_while(move |_| has_base_asset))
    }
```

is way better already. Defining the boolean on line before it's used in `take_while` makes it less painful to understand.

Reply:
Right, that's true. We can avoid both the database read and extra `Box` by doing this:

```rust
    pub fn coins(&self) -> impl Iterator<Item = StorageResult<CoinType>> + '_ {
        let messages = self
            .has_asset(self.base_asset_id)
            .then(|| self.messages_iter());
        self.coins_iter().chain(messages.into_iter().flatten())
    }
```

Reply:
Nice! Updated.